### ISSUE TITLE
Support of Org-mode markup

### ISSUE THREAD

**Issue by tecosaur** [hedgedoc/hedgedoc#1008]
There is currently no (good) way to collaboratively edit [Org-mode](https://orgmode.org/) documents. I would like to change this.

Making something from scratch seems silly, a much better idea would be taking another good plaintext collaborative editing platform and modifying it. This is what I am planning on doing with hedgedoc; I see two ways it could proceed.

-   Hard fork of hedgedoc, rip out the markdown internals and replace them with Org (using [Orga](https://github.com/orgapp/orgajs))
-   Talk to the hedgedoc people, and gauge interest in adding Org support to hedgedoc

As such, I have three questions I'd like to ask:

1.  What do people think of the idea of adding Org support to hedgedoc?
2.  For people who (unlike me [yet]) have delved into the internals of hedgedoc, how much work would you anticipate it being to replace the Markdown components?
3.  For the maintainers — would you be happy receiving questions that arise from an effort to give hedgedoc Org support, and if so where would you like to receive them?

**Comment by davidmehren** [comment]
> What do people think of the idea of adding Org support to hedgedoc?

We (the current developers) have no plans to add more supported markup languages to HedgeDoc. Frankly, it's already complicated enough to build a good editing experience with just markdown.

> how much work would you anticipate it being to replace the Markdown components?

The backend for 2.0 will be mostly markup-agnostic. There will be some markdown-specific things to get things like title and other metadata from a note, but they may be easy enough to replace.  
For the [new frontend](https://github.com/hedgedoc/react-client), you obviously need to replace the renderer. The other components *should* not care too much about the markup, but someone from @hedgedoc/frontend can probably answer that in more detail.

> would you be happy receiving questions that arise from an effort to give hedgedoc Org support, and if so where would you like to receive them?

I'm a bit conflicted here. I do see the appeal of someone taking HedgeDoc, "just" replacing Markdown with their favorite language and not having to worry about all the other things like permissions, real-time communication or login. If these other projects would also contribute back to HedgeDoc, that sounds like an open-source success story!  
But: We are currently in the middle of rewriting HedgeDoc and decided at the start to not support multiple markup languages. This was mostly to keep things simple and the scope manageable. We want to build a markdown editor, not a meta-plug-in-your-own-language editing system.  
When 2.0 is done, and we have a finished, working markdown editor, in my personal opinion, we could look into refactoring things to make it easier for other projects to plug in other languages.

Regarding asking questions: The [development chat](https://matrix.to/#/#hedgedoc-dev:matrix.org?via=shivering-isles.com&via=matrix.org&via=kif.rocks) seems best for now, if it gets too noisy there we can always open another room.

**Comment by mrdrogdrog** [comment]
Regarding frontend:

Yes, most parts are generic, and as long as you keep using yaml frontmatter, you shouldn't experience any problems regarding the meta data.
Yes, the renderer is replaceable. We use slightly different renderers for slide shows, read only view, editor preview, etc.

But, there are parts which are very markdown specific:
- The editor is using markdown highlighting.
- Some of our render components rely on the specific output of the markdown-to-html lib to detect some extensions.
- The sync scrolling relies on an extension in the markdown-to-html lib.
- Some of our code extensions have markdown-like syntax.
- We have some assistant functions like "copy a table from excel, insert it here and it will be inserted as markdown table" or "drop an image here and we'll insert the correct markdown code".
- The TOC is an extension of our markdown-to-html lib
- The cheatsheet and the features page are markdown-specific 

TL;DR: It's not impossible to alter the frontend, but it's not an easy task.

**Comment by tecosaur** [comment]
David, Tilman, thanks for your comments. You've given me a much better idea of what I'm in for.

I'm very happy to hear that the backend is largely format-agnostic, but I see from Tilman's response that (somewhat unsurprisingly) you rely on the specific output format of markdown-to-html lib in many places.

Considering I'm somewhat tight on time at the moment, unless someone else comes along and pushes this I'll probably wait till 2.0 is completed (assuming that happens within the year) before I get going with this. I will doubtlessly have questions, so I'd also like to avoid distracting you from your concerted effort to make v2.0. If, as David raised, at that point there might be any interest in refactoring the internals to make it easier to plug in other languages (and hopefully pull and PR improvements), that would be brilliant.

I appreciate you both being open to a discussion on this :slightly_smiling_face:

**Comment by davidmehren** [comment]
Let's track possible refactoring work in #494 and close this for now.
